src/app/gps/qgsgpscanvasbridge.cpp

```diff
@@ -0,0 +1,436 @@
+#include "qgsprojectdisplaysettings.h"
+#include "qgsbearingnumericformat.h"
+
+QgsGpsCanvasBridge::QgsGpsCanvasBridge( QgsAppGpsConnection *connection, QgsMapCanvas *canvas, QObject *parent )
+  : QObject( parent )
+  , mConnection( connection )
+  , mCanvas( canvas )
+{
+  mWgs84CRS = QgsCoordinateReferenceSystem::fromOgcWmsCrs( QStringLiteral( "EPSG:4326" ) );
+
+  connect( mConnection, &QgsAppGpsConnection::disconnected, this, &QgsGpsCanvasBridge::gpsDisconnected );
+  connect( mConnection, &QgsAppGpsConnection::stateChanged, this, &QgsGpsCanvasBridge::gpsStateChanged );
+
+  connect( QgsGui::instance(), &QgsGui::optionsChanged, this, &QgsGpsCanvasBridge::gpsSettingsChanged );
+
+  mCanvasToWgs84Transform = QgsCoordinateTransform( mCanvas->mapSettings().destinationCrs(), mWgs84CRS, QgsProject::instance() );
+  connect( mCanvas, &QgsMapCanvas::destinationCrsChanged, this, [ = ]
+  {
+    mCanvasToWgs84Transform = QgsCoordinateTransform( mCanvas->mapSettings().destinationCrs(), mWgs84CRS, QgsProject::instance() );
+  } );
+  connect( QgsProject::instance(), &QgsProject::transformContextChanged, this, [ = ]
+  {
+    mCanvasToWgs84Transform = QgsCoordinateTransform( mCanvas->mapSettings().destinationCrs(), mWgs84CRS, QgsProject::instance() );
+  } );
+
+  mDistanceCalculator.setEllipsoid( QgsProject::instance()->ellipsoid() );
+  mDistanceCalculator.setSourceCrs( mWgs84CRS, QgsProject::instance()->transformContext() );
+  connect( QgsProject::instance(), &QgsProject::ellipsoidChanged, this, [ = ]
+  {
+    mDistanceCalculator.setEllipsoid( QgsProject::instance()->ellipsoid() );
+  } );
+
+  connect( mCanvas, &QgsMapCanvas::xyCoordinates, this, &QgsGpsCanvasBridge::cursorCoordinateChanged );
+  connect( mCanvas, &QgsMapCanvas::tapAndHoldGestureOccurred, this, &QgsGpsCanvasBridge::tapAndHold );
+
+  mBearingNumericFormat.reset( QgsLocalDefaultSettings::bearingFormat() );
+  connect( QgsProject::instance()->displaySettings(), &QgsProjectDisplaySettings::bearingFormatChanged, this, [ = ]
+  {
+    mBearingNumericFormat.reset( QgsProject::instance()->displaySettings()->bearingFormat()->clone() );
+    updateGpsDistanceStatusMessage( false );
+  } );
+
+  gpsSettingsChanged();
+
+  mCanvas->installInteractionBlocker( this );
+}
+
+QgsGpsCanvasBridge::~QgsGpsCanvasBridge()
+{
+  delete mMapMarker;
+  mMapMarker = nullptr;
+  delete mMapBearingItem;
+  mMapBearingItem = nullptr;
+
+  if ( mCanvas )
+    mCanvas->removeInteractionBlocker( this );
+}
+
+bool QgsGpsCanvasBridge::blockCanvasInteraction( Interaction interaction ) const
+{
+  switch ( interaction )
+  {
+    case QgsMapCanvasInteractionBlocker::Interaction::MapPanOnSingleClick:
+      // if we're connected and set to follow the GPS location, block the single click navigation mode
+      // to avoid accidental map canvas pans away from the GPS location.
+      // (for now, we don't block click-and-drag pans, as they are less likely to be accidentally triggered)
```

Comment:
One thing that'd be IMHO a good UX: when doing manual panning while the option to auto re-center map is on, it should disable that auto re-center mode.